bpf/kmesh/workload/include/authz.h

```diff
@@ -166,49 +255,146 @@ rule_match_check(Istio__Security__Rule *rule, struct xdp_info *info, struct bpf_
+    return XDP_PASS;
 
+auth_in_user_space:
+    if (bpf_map_delete_elem(&kmesh_tc_args, &tuple_key) != 0) {
+        BPF_LOG(DEBUG, AUTH, "Failed to delete context from map");
+    }
+    bpf_tail_call(ctx, &xdp_tailcall_map, TAIL_CALL_AUTH_IN_USER_SPACE);
+    return XDP_PASS;
+}
+
+SEC("xdp_auth")
+int rule_check(struct xdp_md *ctx)
+{
+    struct match_context *match_ctx;
+    struct bpf_sock_tuple tuple_key = {0};
+    struct xdp_info info = {0};
+    void *rulesPtr;
+    __u64 rule_addr;
+    void *rule;
+    int ret;
+    int i;
+
+    if (construct_tuple_key(ctx, &tuple_key, &info) != PARSER_SUCC) {
+        BPF_LOG(ERR, AUTH, "Failed to get tuple key in rule_check");
+        return XDP_PASS;
+    }
+
+    match_ctx = bpf_map_lookup_elem(&kmesh_tc_args, &tuple_key);
+    if (!match_ctx) {
+        BPF_LOG(ERR, AUTH, "Failed to retrieve match_context from map");
+        return XDP_PASS;
+    }
     for (i = 0; i < MAX_MEMBER_NUM_PER_POLICY; i++) {
-        if (i >= policy->n_rules) {
+        if (i >= match_ctx->n_rules) {
+            BPF_LOG(DEBUG, AUTH, "Rule index %d exceeds rule count %d, exiting loop", i, match_ctx->n_rules);
             break;
         }
-        rule = (Istio__Security__Rule *)kmesh_get_ptr_val((void *)*((__u64 *)rulesPtr + i));
+        if (!match_ctx) {
+            BPF_LOG(ERR, AUTH, "Failed to retrieve match_ctx from map");
+            return XDP_PASS;
+        }
+        rulesPtr = match_ctx->rulesPtr;
+        if (!rulesPtr) {
+            BPF_LOG(ERR, AUTH, "rulesPtr is null");
+            return XDP_PASS;
+        }
+        if (bpf_probe_read_kernel(&rule_addr, sizeof(rule_addr), &rulesPtr[i]) != 0) {
+            BPF_LOG(ERR, AUTH, "Failed to read rule address at index %d", i);
+            continue;
+        }
+
+        rule = (Istio__Security__Rule *)kmesh_get_ptr_val((void *)rule_addr);
         if (!rule) {
             continue;
         }
-        if (rule_match_check(rule, info, tuple_info) == MATCHED) {
-            if (policy->action == ISTIO__SECURITY__ACTION__DENY) {
-                return AUTH_DENY;
-            } else {
-                return AUTH_ALLOW;
+        if (rule_match_check(rule, &info, &tuple_key) == MATCHED) {
+            BPF_LOG(
+                INFO,
+                AUTH,
+                "Rule matched, action: %s",
+                match_ctx->action == ISTIO__SECURITY__ACTION__DENY ? "DENY" : "ALLOW");
+            if (bpf_map_delete_elem(&kmesh_tc_args, &tuple_key) != 0) {
+                BPF_LOG(INFO, AUTH, "Failed to delete tail call context from map");
+            }
+            __u32 auth_result = match_ctx->action == ISTIO__SECURITY__ACTION__DENY ? AUTH_DENY : AUTH_ALLOW;
+            if (bpf_map_update_elem(&map_of_auth, &tuple_key, &auth_result, BPF_ANY) != 0) {
+                BPF_LOG(ERR, AUTH, "Failed to update auth result in map_of_auth");
             }
+            return match_ctx->action == ISTIO__SECURITY__ACTION__DENY ? XDP_DROP : XDP_PASS;
```

Comment:
this logic is not correct. Say we have two authorizationPolicy set, wih the returned result, i think we may miss the second policy check